### PR TITLE
ceph: allow pod labels to be added to CSI components

### LINKED ISSUE #6593: Ceph CSI: Allow custom labels to be added to the Pods
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**

Allow labels to be set on the Pods of the deployed CSI components (e.g., DaemonSets, Deployments).

**What is use case behind this feature:**

Certain monitoring systems rely on certain labels to be set on each Pod in a namespace.

I will create a PR to implement this feature.

### PR DESCRIPTION
**Pull request by galexrt**
**Description of your changes:**

This adds the functionality to add custom pod labels to the CSI
components through the operator configuration way of env vars or config
map.

Resolves #6593

Signed-off-by: Alexander Trost <galexrt@googlemail.com>

**Which issue is resolved by this Pull Request:**
Resolves #6593

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [x] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [x] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [x] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [x] Integration tests have been added, if necessary.
- [x] Pending release notes updated with breaking and/or notable changes, if necessary.
- [x] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [x] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: cluster/charts/rook-ceph/templates/deployment.yaml [hunk 1]
```diff
@@ -160,6 +160,10 @@ spec:
           value: {{ .Values.csi.resizer.image | quote }}
 {{- end }}
 {{- end }}
+{{- if .Values.csi.podLabels }}
+        - name: ROOK_CSI_POD_LABELS
+          value: "{{ .Values.csi.podLabels }}"}}
```
Comment (travisn) [inline cluster/charts/rook-ceph/templates/deployment.yaml:165]: Looks like there are some extra brackets, and also we've been using the `| quote` more often than putting actual quotes around the value, although we certainly have a mix of both approaches.
```suggestion
          value: {{ .Values.csi.podLabels | quote }}
```

File: pkg/operator/k8sutil/labels.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+import (
+	"strings"
+)
+
+// ParseStringToLabels parse a label selector string into a map[string]string
+func ParseStringToLabels(in string) map[string]string {
+	labels := map[string]string{}
+
+	if in == "" {
```
Comment (travisn) [inline pkg/operator/k8sutil/labels.go:27]: This check shouldn't be necessary. The `for` loop will just be a no-op if the string is empty.

File: pkg/operator/k8sutil/labels.go [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+		return labels
+	}
+
+	for _, v := range strings.Split(in, ",") {
+		labelSplit := strings.Split(v, "=")
+
+		// When a value is set
+		if len(labelSplit) > 1 {
```
Comment (travisn) [inline pkg/operator/k8sutil/labels.go:35]: If the length is greater than 2 how about printing a warning even if we don't return an error?

File: cluster/charts/rook-ceph/templates/deployment.yaml [hunk 4]
```diff
@@ -160,6 +160,10 @@ spec:
           value: {{ .Values.csi.resizer.image | quote }}
 {{- end }}
 {{- end }}
+{{- if .Values.csi.podLabels }}
+        - name: ROOK_CSI_POD_LABELS
```
Comment (travisn) [inline cluster/charts/rook-ceph/templates/deployment.yaml:164]: We might want to consider separate labels for the rbd or cephfs drivers, but if there's not a demand for that separation, it's simpler to keep as a single setting.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/k8sutil/labels.go (hunk 2 above)
Comment (galexrt) [reply pkg/operator/k8sutil/labels.go:27]: Interesting, `strings.Split()` always returns at least one element even for an empty string, see https://play.golang.org/p/JmwFMWs7u9a

File: cluster/charts/rook-ceph/templates/deployment.yaml (hunk 4 above)
Comment (galexrt) [reply cluster/charts/rook-ceph/templates/deployment.yaml:164]: Yeah, I thought about that as well. I'll check on the "requirements" I have for this feature, though I think I might go for separate labels env vars for rbd and cephfs CSI components.

File: pkg/operator/k8sutil/labels.go (hunk 2 above)
Comment (travisn) [reply pkg/operator/k8sutil/labels.go:27]: Ah, I was thinking it would be empty, my memory failed me. That really is unexpected to me.

**Review by travisn — APPROVED** [review]

**Pull request merged**